### PR TITLE
[ADD][8.0] PPh 21 Computation on Payslip (combination of #7 #8 #10 #11 #12)

### PR DESCRIPTION
**Pull request by andhit-r**
**This is only for testing purpose only. When all modules good to merge, it all will be split into different PR**

PR List:
1. l10n_id_taxform: https://github.com/OCA/l10n-indonesia/pull/7
2. l10n_id_taxform_pph_21: https://github.com/OCA/l10n-indonesia/pull/8
3. l10n_id_taxform_period: #10 
4. l10n_id_taxform_employee_joining_period: #11 
5. l10n_id_taxform_pph_21_payslip: #12

### CODE REVIEW

**Comment by raditv** [comment]
👍

**Comment by andhit-r** [comment]
@azmimr67 @mikevhe18 please review. Many thanks

**Comment by andhit-r** [comment]
Hi @defcon19, i remember that you have Odoo HR implementation somewhere, i think this PR might come in handy for your project. Would you kindly review it please. Thanks in advance.

**Comment by primeforcindo** [comment]
👍

**Comment by mikevhe18** [comment]
:+1:

**Comment by raditv** [comment]
👍

**Comment by andhit-r** [comment]
Hi @OCA/core-maintainers, l10n-indonesia lacking technical reviewers and PSC vote to review PR. If someone has time to spare would you kindly give technical review to these PRs. Many thanks in advance.

File: l10n_id_taxform_period/models/tax_period.py
Comment (moylop260) [inline l10n_id_taxform_period/models/tax_period.py]: What about use directly the `from openerp import exceptions  exception.ValidationError` instead of a wrapper from models?

File: l10n_id_taxform_period/models/tax_period.py
Comment (moylop260) [inline l10n_id_taxform_period/models/tax_period.py]: same case here

File: l10n_id_taxform_period/models/tax_period.py
Comment (moylop260) [inline l10n_id_taxform_period/models/tax_period.py]: You are using: `results[step - 1]` to return the values.
Maybe you could use `self.search(... offset=step-1, limit=1)` parameter

**Review by moylop260 — APPROVED** [review]
LGTM just small comments
Thanks!

File: l10n_id_taxform_period/models/tax_period.py
Comment (raditv) [reply l10n_id_taxform_period/models/tax_period.py]: 👍

File: l10n_id_taxform_period/models/tax_period.py
Comment (raditv) [reply l10n_id_taxform_period/models/tax_period.py]: Thank you for your suggestion

File: l10n_id_taxform_period/models/tax_period.py
Comment (andhit-r) [reply l10n_id_taxform_period/models/tax_period.py]: Hi @moylop260 . Many thanks for your reviews.

I use wrapper form models because i don't have to import additional module. Does importing from exceptions give technically better result than from models?

Comment (raditv) [reply l10n_id_taxform_period/models/tax_period.py]: I tested it with cProfile, here's the result

```
>>> def test2():
...     raise exception.ValidationError("test2")
... 
>>> cProfile.run('test2()')
         3 function calls in 0.000 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 <stdin>:1(test2)
        1    0.000    0.000    0.000    0.000 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
>>> def test():
...     raise models.ValidationError("error test")
... 
>>> cProfile.run('test()')
         5 function calls in 0.000 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 <stdin>:1(test)
        1    0.000    0.000    0.000    0.000 <string>:1(<module>)
        1    0.000    0.000    0.000    0.000 exceptions.py:33(__init__)
        1    0.000    0.000    0.000    0.000 exceptions.py:68(__init__)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
```

Comment (andhit-r) [reply l10n_id_taxform_period/models/tax_period.py]: Hi @raditv , cprofile obviously way above my league. Could you explain to me the result of cprofile?

Comment (moylop260) [reply l10n_id_taxform_period/models/tax_period.py]: A transitive import is not a good idea because you use more paths to get the main library.
This is similar to have: `from openerp.addons.other_module.models.model import ValidationError`
Or have:

``` python
# file1.py
import os

#file2.py
from file1 import os  # Why not use directly "import os"?
```

Using transitive import requires more caution for new versions.
A example is the deprecated [except_orm](https://github.com/odoo/odoo/blob/dd28f9bf78c35c878eaf35622212e5a40f90fa40/openerp/exceptions.py#L21).

The version of models.py before of deprecating:

``` python
from .exceptions import except_osv
...
```

The version of models.py after deprecating

``` python
from .exceptions import UserError, RedirectWarning
...
```

You old `mymodule.py` with transitive import will be fail `from openerp.models import except_osv` because now is not directly imported from transitive `models.py` file. But if we have a `from openerp.exceptions import except_osv` works perfectly.

We have a similar deprecating case for `exception.warning`  from 8.0 and 9.0 and the lint check this case.
And we have a guideline: https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#idioms
see `Exceptions: Use from openerp.exceptions import Warning as UserError (v8) or from openerp.exceptions import UserError (v9) or find a more appropriate exception in openerp.exceptions.py`

And the lint check this sentences with `openerp.exceptions` but not is checked with `models.exceptions`

Comment (raditv) [reply l10n_id_taxform_period/models/tax_period.py]: @andhit-r there's no significant result because I tested it with high specs server, maybe I need to test it with lower specs. 

But we can see that you need to call 5 functions to use `models.ValidationError`, and for `exception.ValidationError` you only need 3 functions. IMHO `from .exceptions import UserError, RedirectWarning, ValidationError` is better

Comment (andhit-r) [reply l10n_id_taxform_period/models/tax_period.py]: @moylop260 @raditv thank you for your explanation. I will change the exception according suggestion on my next commit.

Comment (andhit-r) [reply l10n_id_taxform_period/models/tax_period.py]: @moylop260 @raditv i have change the way i import exception.

**Comment by moylop260** [comment]
👍

**Comment by andhit-r** [comment]
@raditv @primeforcindo @mikevhe18 @azmimr67 please review the functionality again.

**Comment by azmimr67** [comment]
:+1:

**Comment by andhit-r** [comment]
Hi @OCA/core-maintainers, i will split modules into different PR. My question is, do i need to wait for approval on each PR? Or approval on this PR is enough to merge?

Many thanks.

**Comment by dreispt** [comment]
@andhit-r It's clearer if you follow the normal procedure, and get review approvals for each PR individually. If the review work was already done here, it should be no problem to have the +1 on each PR. If you do have trouble ping us back.

You state that this PR is for testing only. Will you be closing it?

**Comment by andhit-r** [comment]
@dreispt thanks for your explanation. Will wait for approval for each separate PRs. Closing this PR now.

**Pull request closed without merging**
